File: 2022/java/src/main/Day3.java
package main;

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;

public class Day3 {

  public static void main(String[] args) throws IOException {
    AtomicInteger totalItemPriority = new AtomicInteger(0);
    calculatePriorityFromFile("src/resources/day3.txt", line -> {
      var half = line.length() / 2;
      String firstCompartment = line.substring(0, half);
      String secondCompartment = line.substring(half);
      char item = findCommonItem(firstCompartment, secondCompartment);
      totalItemPriority.getAndAdd(calculateCharValue(item));
    });
    System.out.println(totalItemPriority.get());

    AtomicInteger totalBadgePriority = new AtomicInteger(0);
    List<String> lines = new ArrayList<>();
    calculatePriorityFromFile("src/resources/day3.txt", (line) -> {
      lines.add(line);
      if (lines.size() == 3) {
        lines.sort((a, b) -> b.length() - a.length());
        char badge = findCommonBadge(lines.get(0), lines.get(1), lines.get(2));
        totalBadgePriority.getAndAdd(calculateCharValue(badge));
        lines.clear();
      }
    });
    System.out.println(totalBadgePriority.get());
  }

  private static void calculatePriorityFromFile(String filename, Consumer<String> lineConsumer) throws IOException {
    try (BufferedReader br = new BufferedReader(new InputStreamReader(new FileInputStream(filename)))) {
      String line;
      while ((line = br.readLine()) != null) {
        lineConsumer.accept(line);
      }
    }
  }

  private static Character findCommonItem(String firstCompartment, String secondCompartment) {
    for (int i = 0; i < firstCompartment.length(); i++) {
      char c = firstCompartment.charAt(i);
      if (secondCompartment.indexOf(c) > -1)
        return c;
    }
    return '-';
  }

  private static Character findCommonBadge(String firstRucksack, String secondRucksack, String thirdRucksack) {
    for (int i = 0; i < firstRucksack.length(); i++) {
      char c = firstRucksack.charAt(i);
      if (secondRucksack.indexOf(c) > -1 && thirdRucksack.indexOf(c) > -1)
        return c;
    }
    return '-';
  }

  private static int calculateCharValue(char c) {
    if (c >= 'A' && c <= 'Z')
      return (int) c - 38;
    return (int) c - 96;
  }
}

File: 2022/kotlin/src/main/kotlin/Day3.kt
import java.io.BufferedReader
import java.io.FileInputStream
import java.io.InputStreamReader
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Consumer

fun main() {
    val totalItemPriority = AtomicInteger(0)
    calculatePriorityFromFile("src/main/resources/day3.txt") { line: String ->
        val half = line.length / 2
        val firstCompartment = line.substring(0, half)
        val secondCompartment = line.substring(half)
        val item: Char = findCommonItem(firstCompartment, secondCompartment)
        totalItemPriority.getAndAdd(calculateCharValue(item))
    }
    println(totalItemPriority.get())

    val totalBadgePriority = AtomicInteger(0)
    val lines: MutableList<String> = ArrayList()
    calculatePriorityFromFile("src/main/resources/day3.txt") { line: String ->
        lines.add(line)
        if (lines.size == 3) {
            lines.sortedByDescending { it.length }
            val badge: Char = findCommonBadge(lines[0], lines[1], lines[2])
            totalBadgePriority.getAndAdd(calculateCharValue(badge))
            lines.clear()
        }
    }
    println(totalBadgePriority.get())
}

private fun calculatePriorityFromFile(filename: String, lineConsumer: Consumer<String>) {
    BufferedReader(InputStreamReader(FileInputStream(filename))).use { br ->
        var line: String?
        while (br.readLine().also { line = it } != null) {
            lineConsumer.accept(line!!)
        }
    }
}

private fun findCommonItem(firstCompartment: String, secondCompartment: String): Char {
    for (element in firstCompartment) {
        if (secondCompartment.indexOf(element) > -1) return element
    }
    return '-'
}

private fun findCommonBadge(
    firstRucksack: String,
    secondRucksack: String,
    thirdRucksack: String
): Char {
    for (element in firstRucksack) {
        if (secondRucksack.indexOf(element) > -1 && thirdRucksack.indexOf(element) > -1) return element
    }
    return '-'
}

private fun calculateCharValue(c: Char): Int {
    return if (c in 'A'..'Z') c.code - 38 else c.code - 96
}
